README
NAME
    Crypt::MCrypt - Perl interface for libmcrypt C library.

VERSION
    version 0.06

SYNOPSIS
         use Crypt::MCrypt;
 
         my $iv = pack("H*","0000000000000000");
         my $key = pack("H*","1234567890123456" . "7890123456789012" . "1234567890123456");
         my $cipher_text = pack("H*","E9FF3161EE05ABC9" 
             . "7ea3cacb991318aa" 
             . "585379599b0eaabb" 
             . "c4e474ead1956f47" 
             . "6755f13f1af5235d");
         my $algorithm = "tripledes";
         my $mode = "cbc";
         my $obj = Crypt::MCrypt->new(
             algorithm => $algorithm, 
             mode      => $mode,
             key       => $key, 
             iv        => $iv,
         );
         my $plain_text = $obj->decrypt($cipher_text);
         print "\nPLAIN: $plain_text\n";
         print "\nPLAIN in hex: " . unpack("H*",$plain_text) . "\n";
         $cipher_text = $obj->encrypt($plain_text);
         print "\nCIPHER: $cipher_text\n";
         print "\nCIPHER in hex: " . unpack("H*",$cipher_text) . "\n";

DESCRIPTION
    This is a perl interface to libmcrypt c library. It exposes the crypto
    functions provided by the libmcrypt library in a perl interface with a
    binding code that accounts for null C strings in ciphertext or plain
    text.

ATTRIBUTES
  algorithm
    contains the name of the algorithm used to decrypt encrypt blocks of
    data

  mode
    contains the name of the block cipher mode of operation
    <http://en.wikipedia.org/wiki/Block_cipher_modes_of_operation> used to
    encrypt/decrypt blocks of data.

  key
    contains the key for the encryption decryption algorithm

  Initialization Vector
    contains the Initialization for the first block of data in block cipher
    mode of operation.

METHODS
  $self->decrypt($ciphertext)
    decrypt blocks of ciphertext

  $self->encrypt($ciphertext)
    encrypt blocks of data

USAGE
    *   This module provides a object oriented interface to the libmcrypt
        library. It uses Mo, a scaled down version of Moose without any data
        checks to improve speed.

see ALSO
    *   Mo

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through github at
    <https://github.com/shantanubhadoria/crypt-mcrypt/issues>. You will be
    notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/shantanubhadoria/crypt-mcrypt>

      git clone git://github.com/shantanubhadoria/crypt-mcrypt.git

AUTHOR
    Shantanu Bhadoria <shantanu at cpan dott org>

CONTRIBUTORS
    *   Shantanu <shantanu@cpan.org>

    *   Shantanu Bhadoria <shantanu@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Shantanu Bhadoria.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changes for Crypt::MCrypt 

    - Time for the first release


0.05

	- 2013-09-08 17:04:40 Asia/Singapore



0.04

	- 2013-07-11 08:48:57 Asia/Singapore


Adding Tests for DES ECB 

0.03

	- 2013-07-09 23:38:26 Asia/Singapore



Makefile.PL
# This Makefile.PL for  was generated by inc::MCryptMakeMaker .
# Don't edit it but the dist.ini used to construct it.

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl interface for libmcrypt C library.  ",
  "AUTHOR" => "Shantanu Bhadoria <shantanu at cpan dott org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DEFINE" => "",
  "DISTNAME" => "Crypt-MCrypt",
  "EXE_FILES" => [],
  "INC" => "-I.",
  "LIBS" => [
    "-lltdl -lmcrypt "
  ],
  "LICENSE" => "perl",
  "NAME" => "Crypt::MCrypt",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Mo" => 0,
    "XSLoader" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Find" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "List::Util" => 0,
    "Test::More" => 0,
    "Test::Most" => "0.22"
  },
  "VERSION" => "0.06",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw());
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Crypt::MCrypt',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}


lib/Crypt/MCrypt.pm
use strict;
use warnings;

package Crypt::MCrypt;

# PODNAME: Crypt::MCrypt
# ABSTRACT: Perl interface for libmcrypt C library.
#
# This file is part of Crypt-MCrypt
#
# This software is copyright (c) 2013 by Shantanu Bhadoria.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
our $VERSION = '0.06'; # VERSION

use 5.010;

# Dependencies
use Mo;
use Carp;

require XSLoader;
XSLoader::load( 'Crypt::MCrypt', $VERSION );


has algorithm => ();


has mode => ();


has key => ();


has iv => ();


sub decrypt {
    my ( $self, $ciphertext ) = @_;

    my $length = length($ciphertext);
    my $plain_hex =
      Crypt::MCrypt::_decrypt( $self->algorithm, $self->mode, $ciphertext,
        $self->key, $length, $self->iv );
    return pack( "H*", $plain_hex );
}


sub encrypt {
    my ( $self, $plaintext ) = @_;

    my $length = length($plaintext);
    my $cipher_hex =
      Crypt::MCrypt::_encrypt( $self->algorithm, $self->mode, $plaintext,
        $self->key, $length, $self->iv );
    return pack( "H*", $cipher_hex );
}

1;

__END__

=pod

=head1 NAME

Crypt::MCrypt - Perl interface for libmcrypt C library.

=head1 VERSION

version 0.06

=head1 SYNOPSIS

     use Crypt::MCrypt;
 
     my $iv = pack("H*","0000000000000000");
     my $key = pack("H*","1234567890123456" . "7890123456789012" . "1234567890123456");
     my $cipher_text = pack("H*","E9FF3161EE05ABC9" 
         . "7ea3cacb991318aa" 
         . "585379599b0eaabb" 
         . "c4e474ead1956f47" 
         . "6755f13f1af5235d");
     my $algorithm = "tripledes";
     my $mode = "cbc";
     my $obj = Crypt::MCrypt->new(
         algorithm => $algorithm, 
         mode      => $mode,
         key       => $key, 
         iv        => $iv,
     );
     my $plain_text = $obj->decrypt($cipher_text);
     print "\nPLAIN: $plain_text\n";
     print "\nPLAIN in hex: " . unpack("H*",$plain_text) . "\n";
     $cipher_text = $obj->encrypt($plain_text);
     print "\nCIPHER: $cipher_text\n";
     print "\nCIPHER in hex: " . unpack("H*",$cipher_text) . "\n";

=head1 DESCRIPTION

This is a perl interface to libmcrypt c library. It exposes the crypto functions provided by the libmcrypt library in a perl interface 
with a binding code that accounts for null C strings in ciphertext or plain text.

=head1 ATTRIBUTES

=head2 algorithm

contains the name of the algorithm used to decrypt encrypt blocks of data

=head2 mode 

contains the name of the L<block cipher mode of operation|http://en.wikipedia.org/wiki/Block_cipher_modes_of_operation> used to encrypt/decrypt blocks of data.

=head2 key

contains the key for the encryption decryption algorithm

=head2 Initialization Vector 

contains the Initialization for the first block of data in block cipher mode of operation. 

=head1 METHODS

=head2 $self->decrypt($ciphertext)

decrypt blocks of ciphertext 

=head2 $self->encrypt($ciphertext)

encrypt blocks of data

=encoding utf-8

=head1 USAGE

=over

=item *

This module provides a object oriented interface to the libmcrypt library. It uses Mo, a scaled down version of Moose without any data checks to improve speed.

=back

=head1 see ALSO

=over

=item *

L<Mo>

=back

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through github at 
L<https://github.com/shantanubhadoria/crypt-mcrypt/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/shantanubhadoria/crypt-mcrypt>

  git clone git://github.com/shantanubhadoria/crypt-mcrypt.git

=head1 AUTHOR

Shantanu Bhadoria <shantanu at cpan dott org>

=head1 CONTRIBUTORS

=over 4

=item *

Shantanu <shantanu@cpan.org>

=item *

Shantanu Bhadoria <shantanu@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Shantanu Bhadoria.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
